proposals/0011-inline-spirv.md

```diff
@@ -0,0 +1,293 @@
+
+[[SPV_KHR_PostDepthCoverageExecutionMode]]
+float4 PSMain(...) : SV_TARGET
+{ ...
+}
+```
+
+### Instructions and extended instructions
+
+The existing `vk::ext_instruction` attribute is used to define a function as a
+specific SPIR-V instruction. Suppose we wanted to implement the
+[SPV_INTEL_subgroups](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/INTEL/SPV_INTEL_subgroups.html)
+extension in a header file. It contains 8 new instructions. Each one could be
+defined in the header file as a function, and then the function can be called by
+the users. For example,
+
+```
+[[vk::ext_capability(5568)]]
+[[vk::ext_extension("SPV_INTEL_subgroups")]]
+[[vk::ext_instruction(/* OpSubgroupShuffleINTEL */ 5571)]]
+template<typename T>
+T SubgroupShuffleINTEL(T data, uint32 invocationId);
+```
+
+Then the user calls `SubgroupShuffleINTEL()` to use this instruction.
+
+### Types
+
+Some extensions introduce new types. It is possible to define and use a SPIR-V
+type with the existing inline SPIR-V, but it is awkward to use. You can see a
+sample in the
+[spv.intrinsictypeInteger.hlsl](https://github.com/microsoft/DirectXShaderCompiler/blob/128b6fd16b449df696a5c9f9405982903a4f88c4/tools/clang/test/CodeGenSPIRV/spv.intrinsicTypeInteger.hlsl)
+test. Some of the difficulties are:
+
+1.  The definition of the type is split in two. There is a definition of a
+    function which has an arbitrary id and the id for the `OpType*` opcode for
+    the type. The second part calls the function with values for all of the
+    operands to the `OpType*` instruction in SPIR-V.
+1.  The function call must be reachable from the entry point that will use the
+    type even if it will be used to declare a global variable.
+1.  The arbitrary id that is part of the type function is what is used to
+    declare an object of that type. This means the same function cannot be used
+    to declare two different types. Also, if there are multiple header files,
+    they cannot use the same id for two different types. This can become hard to
+    manage.
+
+This proposal deprecates the old mechanism, and replaces it with a new type
+`vk::SpirvType<int OpCode, ...>`. The template on the type contains the opcode
+and all of the parameters necessary for that opcode. The difficulty with this is
+that the operands are not just literal integer values. Sometimes they are
+another type.
+
+The header file could create a partial instantiation with a more meaningful
+name. For example, if you wanted to declare the types from the
+[SPV_INTEL_device_side_avc_motion_estimation](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/INTEL/SPV_INTEL_device_side_avc_motion_estimation.html)
+you could have
+
+```
+template<typename ImageType>
+typedef VmeImageINTEL vk::SpirvType</* OpTypeVmeImageINTEL */ 5700, Imagetype>
+typedef AvcMcePayloadINTEL vk::SprivType</* OpTypeAvcMcePayloadINTEL */ 5704>
```

Comment:
This syntax doesn't really make sense. Probably better to use type alias syntax, given that it's available in HLSL2021, ie
```c++
template <typename ImageType>
using VmeImageINTEL = vk::SpirvType</* OpTypeVmeImageINTEL */ 5700, ImageType>;
```

Reply:
+1 on this, would be amazing, also the templating is very neat